controllers/openstackcluster_controller.go

```diff
@@ -119,16 +120,29 @@ func (r *OpenStackClusterReconciler) Reconcile(ctx context.Context, req ctrl.Req
 
 	// Handle deleted clusters
 	if !openStackCluster.DeletionTimestamp.IsZero() {
-		return reconcileDelete(scope, cluster, openStackCluster)
+		return r.reconcileDelete(ctx, scope, cluster, openStackCluster)
 	}
 
 	// Handle non-deleted clusters
 	return reconcileNormal(scope, cluster, openStackCluster)
 }
 
-func reconcileDelete(scope scope.Scope, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
+func (r *OpenStackClusterReconciler) reconcileDelete(ctx context.Context, scope scope.Scope, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
 	scope.Logger().Info("Reconciling Cluster delete")
 
+	// Wait for machines to be deleted before removing the finalizer as they
+	// depend on this resource to deprovision.  Additionally it appears that
+	// allowing the Kubernetes API to vanish too quickly will upset the capi
+	// kubeadm control plane controller.
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, r.Client, cluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(machines) != 0 {
+		return ctrl.Result{}, fmt.Errorf("waiting for %d machines to be deleted", len(machines))
```

Comment:
Do we want to requeue this immediately? As the machines needs some time to delete it's mayube fair to requeue  after x seconds? I don't have a hard opinion on this.

Reply:
That's a good point! I'm not sure how long it is expected to take to delete the machines but waiting a few seconds (5?) is probably a good idea. So something like this?
```suggestion
		return ctrl.Result{RequeueAfter: 5 * time.Second}, fmt.Errorf("waiting for %d machines to be deleted", len(machines))
```

Reply:
5 seconds is a good value (tbh that‘s exactly the value I wanted to suggest 😄)